include/dlaf/common/assert.h

```diff
@@ -14,102 +14,100 @@
+  std::ostringstream ss;
+  ss << first << '\n' << concat(std::forward<const Ts>(args)...);
+  return ss.str();
+}
 
-#ifdef DLAF_ASSERT_ENABLE
-/// **THIS MACRO MUST BE USED WHEN THE CHECK IS
-/// NEEDED TO ENSURE A CONDITION THAT HAVE VERY LOW IMPACT ON PERFORMANCES.**
+}
+}
+
+/// Each macro has two required parameters and any number of optional parameters:
 ///
-/// Parameters:
-/// 1     condition
-/// 2-*   (optional) comma separated part(s) composing the custom message in case of failure
+/// 1. @Expr        (required) : an experssion that returns a bool
+/// 2. @Msg / @Var  (required) : additional message (even if empty) or variable to print
+/// >2. @Msg / @Var (optional) : optional messages / variables to print
 ///
-/// If the condition is false, it will print an error report and call std::terminate()
+/// If @Expr is false, an error message is composed and `std::terminate()` is called. The error message
+/// contains information about @Expr, the origin of the error and any additional messages / variables provided
+/// as arguments. Any variables passed to the argument needs to be printable, i.e. support `operator<<()`.
 ///
-/// The error report will refer to the given origin and can be extended with a custom message composed
-/// concatenating additional parameters given to the macro.
+/// A disabled ASSERT macro has no overhead, @Expr and any additional arguments are not evaluated.
+/// Macros are enabled/disabled as follows:
 ///
-/// If the switch **DLAF_ASSERT_ENABLE** is not defined, this check will not be performed and it will not
-/// add any overhead, nor for the condition evaluation, nor for the message
-#define DLAF_ASSERT_WITH_ORIGIN(origin, ...) DLAF_CHECK_WITH_ORIGIN("", origin, __VA_ARGS__, '\n')
-
-/// **THIS MACRO MUST BE USED WHEN THE CHECK IS NEEDED TO ENSURE A CONDITION THAT HAVE
-/// VERY LOW IMPACT ON PERFORMANCES.**
+///                       Control flag                 Default
+/// DLAF_ASSERT           DLAF_ASSERT_ENABLE           enabled
+/// DLAF_ASSERT_MODERATE  DLAF_ASSERT_MODERATE_ENABLE  enabled in Debug, disabled in Release
+/// DLAF_ASSERT_HEAVY     DLAF_ASSERT_HEAVY_ENABLE     enabled in Debug, disabled in Release
 ///
-/// Parameters:
-/// 1     condition
-/// 2-*   (optional) comma separated part(s) composing the custom message in case of failure
+/// Examples:
 ///
-/// If the condition is false, it will print an error report and call std::terminate()
+/// ```
+/// bool is_sth_true(...) {...}
+/// std::string my_msg(...) {...}
+/// int a = 3;
+/// int b = 4;
 ///
-/// The error report can be extended with a custom message composed concatenating additional parameters
-/// given to the macro.
+/// DLAF_ASSERT(5 == 6, "");                 // ASSERT with no additional information
+/// DLAF_ASSERT(a == b, "my msg", a, b);     // * ASSERT with additional information and argument
+/// DLAF_ASSERT(is_sth_true(), my_msg());    // ASSERT calling functions
```

Comment:
Please add a note about the fact that if assertions are enabled `my_msg()` is evaluated even if the condition is true.

Reply:
Yes, true this is a downside that with macro we didn't had (it was a requirement in the beginning).

This means that as soon as the assert is enabled, in addition to the fundamental cost of the check, there is always the overhead of building the message with `concat`.

I'm wondering if using a variadic template parameter for `do_assert` cannot be an option, so we can delay the `concat` call inside it and removing the overhead (trading for an additional cost because of the templates at compilation time and in terms of binary size).

Reply:
Good points. The `do_assert()` function now uses variadic template to forward parameters to `concat()`, the function is no longer evaluated if the condition is true. However, `my_msg()` still is, I don't see a way around it unless we use a macro. Do we expect composing messages in functions to be common? If we don't we can leave the code as it is, if we do, then we should use a macro.

Reply:
> However, my_msg() still is, I don't see a way around it unless we use a macro.

Yes, exactly. That was the reason I implemented it with macros in the beginning. Since you already print condition and values (with lazy evaluation) inside the default error message, there is less need of specifying "complex" custom messages, and in the 95% of time is enough just adding a string with a simple message. I think we can overlook this, I would just highlight the "downside" in the documentation as @rasolca suggested.

Reply:
A possibility may be to add an overload to `concat` for functions std::string(void) that evaluates the function at and concatenate it.
Anyway I think it is enough to just document it.